% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
  oneside]{book}
\title{Introduction to multi-omics data analysis}
\author{University of Turku}
\date{2022-01-14}

\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Introduction to multi-omics data analysis},
  pdfauthor={University of Turku},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[top=30mm,left=15mm]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{booktabs}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[]{natbib}
\bibliographystyle{apalike}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{overview}{%
\chapter{Overview}\label{overview}}

\textbf{Welcome to the multi-omics data analysis workshop}

Figure source: Moreno-Indias \emph{et al}. (2021) \href{https://doi.org/10.3389/fmicb.2021.635781}{Statistical and Machine Learning Techniques in Human Microbiome Studies: Contemporary Challenges and Solutions}. Frontiers in Microbiology 12:11.

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

This course is based on \href{https://microbiome.github.io}{\emph{miaverse}} (mia = \textbf{MI}crobiome \textbf{A}nalysis) is an
R/Bioconductor framework for microbiome data science. It extends another popular framework, \href{https://joey711.github.io/phyloseq/}{phyloseq}.

The miaverse consists of an efficient data structure, an
associated package ecosystem, demonstration data sets, and open
documentation. These are explained in more detail in the online book
\href{https://microbiome.github.io/OMA}{Orchestrating Microbiome Analysis}.

This workshop material walks you through example workflows for multi-omics data
analysis covering data access, exploration, analysis, visualization and reproducible
reporting. \textbf{You can run the workflow by simply copy-pasting the
examples.} For advanced material, you can test and modify further
examples from the \href{https://microbiome.github.io/OMA}{OMA book}, or try
to apply the techniques to your own data.

\hypertarget{learning-goals}{%
\section{Learning goals}\label{learning-goals}}

This workshop provides an overview of analytical
tools for multi-omics studies in R. A particular focus is on multi-omics tools and techniques
required to process microbial community data in combination with other omics.

After the workshop the participants should be able to preprocess and manipulate data,
perform simple visualizations and statistical analyses, apply unsupervised and supervised
machine learning, and produce robust and reproducible results.

\textbf{Target audience}
Advanced students and applied researchers who wish
to develop their skills in multi-omics analysis.

\textbf{Venue}
The course is organized fully remotely in \href{https://utu.zoom.us/j/63432425453}{Zoom}.
The meeting requires a passcode (sent by email to the participants).

\hypertarget{acknowledgments}{%
\section{Acknowledgments}\label{acknowledgments}}

\textbf{Citation} ``Introduction to microbiome data science (2021). URL: \url{https://microbiome.github.io}''.

\citet{FindingPheno2022workshop}

We thank Felix Ernst, Sudarshan Shetty, and other \href{https://microbiome.github.io}{miaverse
developers} who have contributed open
resources that supported the development of the training material.

\textbf{Contact} \href{http://datascience.utu.fi}{Leo Lahti}, University of Turku, Finland

\textbf{License} All material is released under the open \href{LICENSE}{CC BY-NC-SA 3.0 License}.

\textbf{Source code}

The source code of this repository is fully reproducible and contains
the Rmd files with executable code. All files can be rendered at one
go by running the file \url{main.R}. You can check the file for
details on how to clone the repository and convert it into a gitbook,
although this is not necessary for the training.

\begin{itemize}
\tightlist
\item
  Landing page (html): \href{https://microbiome.github.io/course_2022_FindingPheno/}{workshop teaching material}
\item
  Source code (github): \href{https://github.com/microbiome/course_2022_FindingPheno}{workshop teaching material}
\end{itemize}

\hypertarget{program}{%
\chapter{Program}\label{program}}

The workshop takes place on the 13th and 14th of January from 9am -- 5pm
(CET). Short breaks will be scheduled between sessions.

The practical sessions consists of a set of example
multi-omics analysis workflows. It is assumed that you
have already installed the required software. Do not hesitate to ask
support from the course assistants.

\hypertarget{day-1}{%
\section{Day 1}\label{day-1}}

\textbf{Lectures}

\begin{itemize}
\item
  9:15-10:00 - Welcome and introduction - Leo Lahti, Associate professor (UTU)
\item
  10:00-10:15 - Break
\item
  10:15-11:00 - Metagenomics - Katariina Pärnänen, Postdoctoral researcher (UTU)
\item
  11:00-11:15 - Break
\item
  11:15-12:00 - Metabolomics - Pande Putu Erawijantari, Postdoctoral researcher (UTU)
\item
  12:00-12:15 - Break
\item
  12:15-13:00 - Multiomics - Leo Lahti, Associate professor (UTU)
\item
  13:00-14:15 - Lunch break
\end{itemize}

\textbf{Practical}

\begin{itemize}
\item
  14:15-17:00 - Tuomas Borman and Chouaib Benchraka, Research assistants (UTU)
\item
  Topics:

  \begin{itemize}
  \item
    Data import and data structures
  \item
    Microbiome data exploration
  \item
    Visualization
  \end{itemize}
\item
  R Code:

  \begin{itemize}
  \item
    \href{install_script.R}{Installation}
  \item
    \href{Day_1.R}{Day 1}
  \end{itemize}
\end{itemize}

\begin{center}\rule{0.5\linewidth}{0.5pt}\end{center}

\hypertarget{day-2}{%
\section{Day 2}\label{day-2}}

\textbf{Lectures}

\begin{itemize}
\item
  9:15-11:00 (including a short break) - Unsupervised and supervised machine learning - Matti Ruuskanen, Postdoctoral researcher (UTU)
\item
  11:00-11:15 - Break
\item
  11:15-12:00 - Individual-based modeling - Gergely Boza, Research fellow (CER)
\item
  12:00-12:15 - Break
\item
  12:15-13:00 - Data integration - Leo Lahti, Associate professor (UTU)
\item
  13:00-14:15 - Lunch break
\end{itemize}

\textbf{Practical}

\begin{itemize}
\item
  14:15-17:00 - Tuomas Borman, Matti Ruuskanen and Chouaib Benchraka (UTU)
\item
  Topics:

  \begin{itemize}
  \item
    Unsupervised learning: Beta-diversity and biclustering
  \item
    Supervised learning: Regression and classification with random forests
  \item
    Validation and interpretation of black box models
  \end{itemize}
\item
  R Code:

  \begin{itemize}
  \item
    \href{install_script.R}{Installation}
  \item
    \href{Day_2.R}{Day 2}
  \end{itemize}
\end{itemize}

\hypertarget{getting-started}{%
\chapter{Getting started}\label{getting-started}}

\hypertarget{checklist-before-the-workshop}{%
\section{Checklist (before the workshop)}\label{checklist-before-the-workshop}}

Install the following software in advance in order to avoid
unnecessary delays and leaving more time for the workshop contents.

\begin{itemize}
\item
  \href{https://www.r-project.org/}{R (version \textgreater4.1.0)}
\item
  \href{https://www.rstudio.com/products/rstudio/download/}{RStudio};
  choose ``Rstudio Desktop'' to download the latest version. Optional
  but preferred. For further details, check the \href{https://www.rstudio.com/}{Rstudio home
  page}.
\item
  For Windows users: \href{https://cran.r-project.org/bin/windows/Rtools/rtools40.html}{Rtools};
  Follow the instructions to install the toolkit. This might be required to compile some of the
  packages required on this course.
\item
  Install and load the required R packages
\end{itemize}

\hypertarget{support-and-resources}{%
\section{Support and resources}\label{support-and-resources}}

For online support on installation and other matters, you can join us at:

\begin{itemize}
\tightlist
\item
  Users: \href{https://gitter.im/microbiome/miaverse?utm_source=badge\&utm_medium=badge\&utm_campaign=pr-badge\&utm_content=badge}{miaverse Gitter channel}
\item
  Developers: \href{https://bioc-community.herokuapp.com}{Bioconductor Slack} \#microbiomeexperiment channel (ask for an invitation)
\end{itemize}

\hypertarget{installing-and-loading-the-required-r-packages}{%
\section{Installing and loading the required R packages}\label{installing-and-loading-the-required-r-packages}}

This section shows how to install and load all required packages into
the R session. Only uninstalled packages are installed.

\begin{Shaded}
\begin{Highlighting}[]
\ControlFlowTok{if}\NormalTok{ (}\SpecialCharTok{!}\FunctionTok{requireNamespace}\NormalTok{(}\StringTok{"BiocManager"}\NormalTok{)) \{}
  \FunctionTok{install.packages}\NormalTok{(}\StringTok{"BiocManager"}\NormalTok{)}
\NormalTok{\}}

\CommentTok{\# List of packages that we need from cran and bioc }
\NormalTok{packages }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"ggplot2"}\NormalTok{, }\StringTok{"pheatmap"}\NormalTok{, }\StringTok{"stringr"}\NormalTok{, }\StringTok{"igraph"}\NormalTok{, }\StringTok{"ANCOMBC"}\NormalTok{,}
             \StringTok{"microbiome"}\NormalTok{, }\StringTok{"httpuv"}\NormalTok{, }\StringTok{"microbiomeDataSets"}\NormalTok{, }\StringTok{"mia"}\NormalTok{, }\StringTok{"caret"}\NormalTok{, }\StringTok{"ranger"}\NormalTok{,}
            \StringTok{"dplyr"}\NormalTok{, }\StringTok{"miaViz"}\NormalTok{, }\StringTok{"knitr"}\NormalTok{, }\StringTok{"kableExtra"}\NormalTok{, }\StringTok{"vegan"}\NormalTok{, }\StringTok{"ecodist"}\NormalTok{, }\StringTok{"biclust"}\NormalTok{,}
            \StringTok{"patchwork"}\NormalTok{, }\StringTok{"pdp"}\NormalTok{, }\StringTok{"MLmetrics"}\NormalTok{, }\StringTok{"precrec"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The following script tries to load all required packages, and if they are not available, installs them.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{new.pkgs }\OtherTok{\textless{}{-}}\NormalTok{ packages[}\SpecialCharTok{!}\NormalTok{(packages }\SpecialCharTok{\%in\%} \FunctionTok{installed.packages}\NormalTok{()[, }\StringTok{"Package"}\NormalTok{])]}

\ControlFlowTok{if}\NormalTok{ (}\FunctionTok{length}\NormalTok{(new.pkgs)) \{}
\NormalTok{    BiocManager}\SpecialCharTok{::}\FunctionTok{install}\NormalTok{(new.pkgs, }\AttributeTok{ask=}\NormalTok{T)}
\NormalTok{\}}

\FunctionTok{sapply}\NormalTok{(packages, require, }\AttributeTok{character.only =} \ConstantTok{TRUE}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##            ggplot2           pheatmap            stringr             igraph 
##               TRUE               TRUE               TRUE               TRUE 
##            ANCOMBC         microbiome             httpuv microbiomeDataSets 
##               TRUE               TRUE               TRUE               TRUE 
##                mia              caret             ranger              dplyr 
##               TRUE               TRUE               TRUE               TRUE 
##             miaViz              knitr         kableExtra              vegan 
##               TRUE               TRUE               TRUE               TRUE 
##            ecodist            biclust          patchwork                pdp 
##               TRUE               TRUE               TRUE               TRUE 
##          MLmetrics            precrec 
##               TRUE               TRUE
\end{verbatim}

\hypertarget{data}{%
\chapter{Data}\label{data}}

This section demonstrates how to import data in R.

\hypertarget{data-structure}{%
\section{Data structure}\label{data-structure}}

Such analysis using the miaverse framework, are based upon core data structures
including SingleCellExperiment (SCE), SummarizedExperiment (SE), TreeSummarizedExperiment (TreeSE) and MultiAssayExperiment (MAE) \href{https://microbiome.github.io/course_2022_miaverse/study-material.html\#resources-for-treesummarizedexperiment}{(resources)}.

Multi-assay data can be stored in \href{https://microbiome.github.io/OMA/containers.html\#alternative-experiments}{altExp}
slot of TreeSE or \href{https://microbiome.github.io/OMA/containers.html\#multiassayexperiments}{MAE} data container.

Different data sets are first imported into SE or TreeSE data container similarly to the case when only one data set is present. After that different data sets are combined into the same data container. Result is one TreeSE object with alternative experiment in altExp slot, or MAE object with multiple experiment in its experiment slot.

\hypertarget{example-data}{%
\section{Example data}\label{example-data}}

As an example data, we use data from following publication: Hintikka L et al.~(2021) \href{https://www.mdpi.com/1660-4601/18/8/4049}{Xylo-oligosaccharides in prevention of hepatic steatosis and adipose tissue inflammation: associating taxonomic and metabolomic patterns in fecal microbiotas with biclustering}.

This example data can be loaded from \href{https://bioconductor.org/packages/release/data/experiment/html/microbiomeDataSets.html}{microbiomeDataSets}. The data is already in MAE format. It includes three different experiments: microbial abundance data, metabolite concentrations, and data about different biomarkers.

\hypertarget{importing-data-in-r}{%
\section{Importing data in R}\label{importing-data-in-r}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(stringr)}

\CommentTok{\# Load the data}
\NormalTok{mae }\OtherTok{\textless{}{-}}\NormalTok{ microbiomeDataSets}\SpecialCharTok{::}\FunctionTok{HintikkaXOData}\NormalTok{()}

\CommentTok{\# Drop off those bacteria that do not include information in Phylum or lower levels}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}}\NormalTok{ mae[[}\DecValTok{1}\NormalTok{]][}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{Phylum), ]}

\CommentTok{\# Clean taxonomy data, so that names do not include additional characters}
\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]) }\OtherTok{\textless{}{-}} \FunctionTok{DataFrame}\NormalTok{(}\FunctionTok{apply}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]), }\DecValTok{2}\NormalTok{, }
\NormalTok{                                     str\_remove, }\AttributeTok{pattern =} \StringTok{".\_[0{-}9]\_\_"}\NormalTok{))}

\NormalTok{mae}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## A MultiAssayExperiment object of 3 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 3:
##  [1] microbiota: SummarizedExperiment with 12613 rows and 40 columns
##  [2] metabolites: SummarizedExperiment with 38 rows and 40 columns
##  [3] biomarkers: SummarizedExperiment with 39 rows and 40 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save data to flat files
\end{verbatim}

\hypertarget{microbiome-data-exploration}{%
\chapter{Microbiome data exploration}\label{microbiome-data-exploration}}

Now we have loaded the data set into R. Next, let us walk through some
basic operations for data exploration to confirm that the data has all
the necessary components.

\hypertarget{data-structure-1}{%
\section{Data structure}\label{data-structure-1}}

Let us now investigate how taxonomic profiling data is organized in R.

Dimensionality tells us how many taxa and samples the data
contains. As we can see, there are 12613 taxa and 40
samples.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# mae[[1]]: indexing/retrieving the taxonomic data experiment}
\FunctionTok{dim}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12613    40
\end{verbatim}

The \texttt{rowData} slot contains a taxonomic table. This includes taxonomic
information for each of the 12613 entries. With the \texttt{head()}
command, we can print just the beginning of the table.

The \texttt{rowData} seems to contain information from 7
different taxonomy classes.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(}\FunctionTok{head}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]]))) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{kable\_styling}\NormalTok{(}\StringTok{"striped"}\NormalTok{, }
                            \AttributeTok{latex\_options=}\StringTok{"scale\_down"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{scroll\_box}\NormalTok{(}\AttributeTok{width =} \StringTok{"100\%"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|l|l|l|l|l|l|l}
\hline
  & Phylum & Class & Order & Family & Genus & Species & OTU\\
\hline
GAYR01026362.62.2014 & Proteobacteria & Alphaproteobacteria & Rickettsiales & Mitochondria & Solanum melongena (eggplant) & Solanum melongena (eggplant) & GAYR01026362.62.2014\\
\hline
CVJT01000011.50.2173 & Firmicutes & Bacilli & Bacillales & Staphylococcaceae & Staphylococcus & Staphylococcus aureus & CVJT01000011.50.2173\\
\hline
KF625183.1.1786 & Proteobacteria & Gammaproteobacteria & Enterobacteriales & Enterobacteriaceae & Klebsiella & Klebsiella oxytoca & KF625183.1.1786\\
\hline
AYSG01000002.292.2076 & Firmicutes & Bacilli & Lactobacillales & Streptococcaceae & Streptococcus & Streptococcus thermophilus TH1435 & AYSG01000002.292.2076\\
\hline
CCPS01000022.154.1916 & Proteobacteria & Gammaproteobacteria & Enterobacteriales & Enterobacteriaceae & Escherichia-Shigella & Escherichia coli & CCPS01000022.154.1916\\
\hline
KJ923794.1.1762 & Firmicutes & Bacilli & Bacillales & Staphylococcaceae & Staphylococcus & Staphylococcus aureus & KJ923794.1.1762\\
\hline
\end{tabular}}
\end{table}

The colData slot contains sample metadata. It contains information for all 40 samples.
However, here only the 6 first samples are shown as we use the \texttt{head()} command. There
are 6 columns, that contain information, e.g., about patients' status, and cohort.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# For simplicity, classify all high{-}fat diets as high{-}fat, and all the low{-}fat }
\CommentTok{\# diets as low{-}fat diets}
\FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\SpecialCharTok{==} \StringTok{"High{-}fat"} \SpecialCharTok{|} 
                              \FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet }\SpecialCharTok{==} \StringTok{"High{-}fat + XOS"}\NormalTok{, }
                            \StringTok{"High{-}fat"}\NormalTok{, }\StringTok{"Low{-}fat"}\NormalTok{)}

\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(}\FunctionTok{head}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae))) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{kable\_styling}\NormalTok{(}\StringTok{"striped"}\NormalTok{, }
                            \AttributeTok{latex\_options=}\StringTok{"scale\_down"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{scroll\_box}\NormalTok{(}\AttributeTok{width =} \StringTok{"100\%"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|l|l|l|l|l|r}
\hline
  & Sample & Rat & Site & Diet & Fat & XOS\\
\hline
C1 & C1 & 1 & Cecum & High-fat & High & 0\\
\hline
C2 & C2 & 2 & Cecum & High-fat & High & 0\\
\hline
C3 & C3 & 3 & Cecum & High-fat & High & 0\\
\hline
C4 & C4 & 4 & Cecum & High-fat & High & 0\\
\hline
C5 & C5 & 5 & Cecum & High-fat & High & 0\\
\hline
C6 & C6 & 6 & Cecum & High-fat & High & 0\\
\hline
\end{tabular}}
\end{table}

From here, we can draw summaries of the sample (column) data, for
instance to see what is the diet distribution.

The command \texttt{colData(mae)\$Diet} fetches the data from the
column, and \texttt{table()} creates a table that shows how many times each
class is present, and \texttt{sort()} sorts the table to ascending order.

There are 20
samples from mice having High-fat,
and 20 Low-fat.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{sort}\NormalTok{(}\FunctionTok{table}\NormalTok{(}\FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## High-fat  Low-fat 
##       20       20
\end{verbatim}

\hypertarget{transformations}{%
\subsection{Transformations}\label{transformations}}

Microbial abundances are typically `compositional' (relative) in the
current microbiome profiling data sets. This is due to technical
aspects of the data generation process (see e.g.~\href{https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full}{Gloor et al.,
2017}).

The next example calculates relative abundances as these are usually easier to
interpret than plain counts. For some statistical models we need to
transform the data into other formats as explained in above link (and
as we will see later).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Calculates relative abundances, and stores the table to assays}
\NormalTok{mae[[}\DecValTok{1}\NormalTok{]] }\OtherTok{\textless{}{-}} \FunctionTok{transformCounts}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{method =} \StringTok{"relabundance"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

A variety of standard transformations for microbiome data are available through \href{https://microbiome.github.io/mia/reference/transformCounts.html}{mia R package}.

\hypertarget{aggregation}{%
\subsection{Aggregation}\label{aggregation}}

Microbial species can be called at multiple taxonomic resolutions. We
can easily agglomerate the data based on taxonomic ranks. Here, we
agglomerate the data at Phylum level.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{se\_phylum }\OtherTok{\textless{}{-}} \FunctionTok{agglomerateByRank}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{)}

\CommentTok{\# Show dimensionality}
\FunctionTok{dim}\NormalTok{(se\_phylum)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 13 40
\end{verbatim}

Now there are 13 taxa and 40
samples, meaning that there are 13 different
Phylum level taxonomic groups. Looking at the \texttt{rowData} after
agglomeration shows all Firmicutes are combined together, and all
lower rank information is lost.

From the assay we can see that all abundances of taxa that belong to
Firmicutes are summed up.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(}\FunctionTok{head}\NormalTok{(}\FunctionTok{rowData}\NormalTok{(se\_phylum))) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{kable\_styling}\NormalTok{(}\StringTok{"striped"}\NormalTok{, }
                            \AttributeTok{latex\_options=}\StringTok{"scale\_down"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{scroll\_box}\NormalTok{(}\AttributeTok{width =} \StringTok{"100\%"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|l|l|l|l|l|l|l}
\hline
  & Phylum & Class & Order & Family & Genus & Species & OTU\\
\hline
Proteobacteria & Proteobacteria & NA & NA & NA & NA & NA & GAYR01026362.62.2014\\
\hline
Firmicutes & Firmicutes & NA & NA & NA & NA & NA & CVJT01000011.50.2173\\
\hline
Cyanobacteria & Cyanobacteria & NA & NA & NA & NA & NA & GEMN01027092.33.1623\\
\hline
Tenericutes & Tenericutes & NA & NA & NA & NA & NA & AM277369.1.1548\\
\hline
Deferribacteres & Deferribacteres & NA & NA & NA & NA & NA & AYGZ01000001.327.1863\\
\hline
Actinobacteria & Actinobacteria & NA & NA & NA & NA & NA & JGZF01000005.1.1534\\
\hline
\end{tabular}}
\end{table}

If you are sharp, you have by now noticed that all the aggregated
values in the above example are NA's (missing data). This is because
the agglomeration is missing abundances for certain taxa, and in that
case the sum is not defined by default (\texttt{na.rm\ =\ FALSE}). We can
ignore the missing values in summing up the data by setting \texttt{na.rm\ =\ TRUE}; then the taxa that do not have information in specified level
will be removed. Those taxa that do not have information in specified
level are agglomerated at lowest possible level that is left after
agglomeration.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{temp }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(}\FunctionTok{agglomerateByRank}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{))}

\CommentTok{\# Prints those taxa that do not have information at the Genus level (NA)}
\NormalTok{knitr}\SpecialCharTok{::}\FunctionTok{kable}\NormalTok{(}\FunctionTok{head}\NormalTok{(temp[}\FunctionTok{which}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(temp}\SpecialCharTok{$}\NormalTok{Genus)),])) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{kable\_styling}\NormalTok{(}\StringTok{"striped"}\NormalTok{, }
                            \AttributeTok{latex\_options=}\StringTok{"scale\_down"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
\NormalTok{  kableExtra}\SpecialCharTok{::}\FunctionTok{scroll\_box}\NormalTok{(}\AttributeTok{width =} \StringTok{"100\%"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{table}
\centering
\resizebox{\linewidth}{!}{
\begin{tabular}{l|l|l|l|l|l|l|l}
\hline
  & Phylum & Class & Order & Family & Genus & Species & OTU\\
\hline
Family:uncultured & Proteobacteria & Alphaproteobacteria & Rhodospirillales & uncultured & NA & NA & JRJTB:01000:00983\\
\hline
Family:Ruminococcaceae & Firmicutes & Clostridia & Clostridiales & Ruminococcaceae & NA & NA & JRJTB:00751:00256\\
\hline
Order:Clostridiales & Firmicutes & Clostridia & Clostridiales & NA & NA & NA & JRJTB:03059:01977\\
\hline
Family:Lachnospiraceae & Firmicutes & Clostridia & Clostridiales & Lachnospiraceae & NA & NA & JRJTB:00738:02832\\
\hline
Family:Peptostreptococcaceae & Firmicutes & Clostridia & Clostridiales & Peptostreptococcaceae & NA & NA & JRJTB:01731:00274\\
\hline
Family:Pasteurellaceae & Proteobacteria & Gammaproteobacteria & Pasteurellales & Pasteurellaceae & NA & NA & JRJTB:01960:01703\\
\hline
\end{tabular}}
\end{table}

Here agglomeration is done similarly, but na.rm = TRUE

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{temp2 }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(}\FunctionTok{agglomerateByRank}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{rank =} \StringTok{"Genus"}\NormalTok{, }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{))}

\FunctionTok{print}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{"Agglomeration with na.rm = FALSE: "}\NormalTok{, }\FunctionTok{dim}\NormalTok{(temp)[}\DecValTok{1}\NormalTok{], }\StringTok{" taxa."}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Agglomeration with na.rm = FALSE: 277 taxa."
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{"Agglomeration with na.rm = TRUE: "}\NormalTok{, }\FunctionTok{dim}\NormalTok{(temp2)[}\DecValTok{1}\NormalTok{], }\StringTok{" taxa."}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Agglomeration with na.rm = TRUE: 262 taxa."
\end{verbatim}

The \href{https://microbiome.github.io/mia/reference/index.html}{mia
package}
contains further examples on various data agglomeration and splitting
options.

\hypertarget{visualization}{%
\section{Visualization}\label{visualization}}

The \href{https://microbiome.github.io/miaViz/}{miaViz package} facilitates
data visualization. Let us plot the Phylum level abundances.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Here we specify "relabundance" to be abundance table that we use for plotting.}
\CommentTok{\# Note that we can use agglomerated or non{-}agglomerated mae[[1]] as an input, because}
\CommentTok{\# the function agglomeration is built{-}in option. }

\CommentTok{\# Legend does not fit into picture, so its height is reduced.}
\NormalTok{plot\_abundance }\OtherTok{\textless{}{-}} \FunctionTok{plotAbundance}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]], }\AttributeTok{abund\_values=}\StringTok{"relabundance"}\NormalTok{, }\AttributeTok{rank =} \StringTok{"Phylum"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{legend.key.height =} \FunctionTok{unit}\NormalTok{(}\FloatTok{0.5}\NormalTok{, }\StringTok{"cm"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{scale\_y\_continuous}\NormalTok{(}\AttributeTok{label =}\NormalTok{ scales}\SpecialCharTok{::}\NormalTok{percent)}

\NormalTok{plot\_abundance }
\end{Highlighting}
\end{Shaded}

\includegraphics{05-explore_files/figure-latex/unnamed-chunk-10-1.pdf}

\textbf{Density plot} shows the overall abundance distribution for a given
taxonomic group. Let us check the relative abundance of Firmicutes
across the sample collection. The density plot is a smoothened
version of a standard histogram.

The plot shows peak abundances around 30 \%.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Subset data by taking only Firmicutes}
\NormalTok{se\_firmicutes }\OtherTok{\textless{}{-}}\NormalTok{ se\_phylum[}\StringTok{"Firmicutes"}\NormalTok{]}

\CommentTok{\# Gets the abundance table}
\NormalTok{abundance\_firmicutes }\OtherTok{\textless{}{-}} \FunctionTok{assay}\NormalTok{(se\_firmicutes, }\StringTok{"relabundance"}\NormalTok{)}

\CommentTok{\# Creates a data frame object, where first column includes abundances}
\NormalTok{firmicutes\_abund\_df }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(}\FunctionTok{t}\NormalTok{(abundance\_firmicutes))}
\CommentTok{\# Rename the first and only column}
\FunctionTok{colnames}\NormalTok{(firmicutes\_abund\_df) }\OtherTok{\textless{}{-}} \StringTok{"abund"}

\CommentTok{\# Creates a plot. Parameters inside feom\_density are optional. With }
\CommentTok{\# geom\_density(bw=1000), it is possible to adjust bandwidth.}
\NormalTok{firmicutes\_abund\_plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(firmicutes\_abund\_df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ abund)) }\SpecialCharTok{+} 
  \FunctionTok{geom\_density}\NormalTok{(}\AttributeTok{color=}\StringTok{"darkred"}\NormalTok{, }\AttributeTok{fill=}\StringTok{"lightblue"}\NormalTok{) }\SpecialCharTok{+} 
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"Relative abundance"}\NormalTok{, }\AttributeTok{title =} \StringTok{"Firmicutes"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_classic}\NormalTok{() }\SpecialCharTok{+} \CommentTok{\# Changes the background}
  \FunctionTok{scale\_x\_continuous}\NormalTok{(}\AttributeTok{label =}\NormalTok{ scales}\SpecialCharTok{::}\NormalTok{percent)}

\NormalTok{firmicutes\_abund\_plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{05-explore_files/figure-latex/unnamed-chunk-11-1.pdf}

For more visualization options and examples, see the \href{https://microbiome.github.io/miaViz/articles/miaViz.html}{miaViz vignette}.

\hypertarget{beta-diversity}{%
\chapter{Beta diversity}\label{beta-diversity}}

Beta diversity is another name for sample dissimilarity. It quantifies
differences in the overall taxonomic composition between two samples.

Common indices include Bray-Curtis, Unifrac, Jaccard index, and the
Aitchison distance. For more background information
and examples, you can check the dedicated section in \href{https://microbiome.github.io/OMA/microbiome-diversity.html\#beta-diversity}{online
book}.

\hypertarget{examples-of-pcoa-with-different-settings}{%
\section{Examples of PCoA with different settings}\label{examples-of-pcoa-with-different-settings}}

Beta diversity estimation generates a (dis)similarity matrix that
contains for each sample (rows) the dissimilarity to any other sample
(columns).

This complex set of pairwise relations can be visualized in
informative ways, and even coupled with other explanatory
variables. As a first step, we compress the information to a lower
dimensionality, or fewer principal components, and then visualize
sample similarity based on that using ordination techniques, such as
Principal Coordinate Analysis (PCoA). PCoA is a non-linear dimension
reduction technique, and with Euclidean distances it is is identical
to the linear PCA (except for potential scaling).

We typically retain just the two (or three) most informative top
components, and ignore the other information. Each sample has a score
on each of these components, and each component measures the variation
across a set of correlated taxa. The top components are then easily
visualized on a two (or three) dimensional display.

Let us next look at some concrete examples.

\hypertarget{pcoa-for-asv-level-data-with-bray-curtis}{%
\subsection{PCoA for ASV-level data with Bray-Curtis}\label{pcoa-for-asv-level-data-with-bray-curtis}}

Let us start with PCoA based on a Bray-Curtis dissimilarity matrix.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Pick the relative abundance table}
\NormalTok{rel\_abund\_assay }\OtherTok{\textless{}{-}} \FunctionTok{assays}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{relabundance}

\CommentTok{\# Calculates Bray{-}Curtis distances between samples. Because taxa is in}
\CommentTok{\# columns, it is used to compare different samples. We transpose the}
\CommentTok{\# assay to get taxa to columns}
\NormalTok{bray\_curtis\_dist }\OtherTok{\textless{}{-}}\NormalTok{ vegan}\SpecialCharTok{::}\FunctionTok{vegdist}\NormalTok{(}\FunctionTok{t}\NormalTok{(rel\_abund\_assay), }\AttributeTok{method =} \StringTok{"bray"}\NormalTok{)}

\CommentTok{\# PCoA}
\NormalTok{bray\_curtis\_pcoa }\OtherTok{\textless{}{-}}\NormalTok{ ecodist}\SpecialCharTok{::}\FunctionTok{pco}\NormalTok{(bray\_curtis\_dist)}

\CommentTok{\# All components could be found here: }
\CommentTok{\# bray\_curtis\_pcoa$vectors}
\CommentTok{\# But we only need the first two to demonstrate what we can do:}
\NormalTok{bray\_curtis\_pcoa\_df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{pcoa1 =}\NormalTok{ bray\_curtis\_pcoa}\SpecialCharTok{$}\NormalTok{vectors[,}\DecValTok{1}\NormalTok{], }
                                  \AttributeTok{pcoa2 =}\NormalTok{ bray\_curtis\_pcoa}\SpecialCharTok{$}\NormalTok{vectors[,}\DecValTok{2}\NormalTok{])}

\CommentTok{\# Create a plot}
\NormalTok{bray\_curtis\_plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ bray\_curtis\_pcoa\_df, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x=}\NormalTok{pcoa1, }\AttributeTok{y=}\NormalTok{pcoa2)) }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"PC1"}\NormalTok{,}
       \AttributeTok{y =} \StringTok{"PC2"}\NormalTok{, }
       \AttributeTok{title =} \StringTok{"Bray{-}Curtis PCoA"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{(}\DecValTok{12}\NormalTok{) }\CommentTok{\# makes titles smaller}

\NormalTok{bray\_curtis\_plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{06-beta_files/figure-latex/pcoa_asv_bc-1.pdf}

\hypertarget{highlighting-external-variables}{%
\section{Highlighting external variables}\label{highlighting-external-variables}}

We can map other variables on the same plot for example by coloring
the points accordingly.

The following is an example with a discrete grouping variable (Diet) shown with colors:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Add diet information to data.frame}
\NormalTok{bray\_curtis\_pcoa\_df}\SpecialCharTok{$}\NormalTok{Diet }\OtherTok{\textless{}{-}} \FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet}

\CommentTok{\# Creates a plot}
\NormalTok{plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ bray\_curtis\_pcoa\_df, }\FunctionTok{aes\_string}\NormalTok{(}\AttributeTok{x =} \StringTok{"pcoa1"}\NormalTok{, }\AttributeTok{y =} \StringTok{"pcoa2"}\NormalTok{, }\AttributeTok{color =} \StringTok{"Diet"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x =} \StringTok{"PC1"}\NormalTok{,}
       \AttributeTok{y =} \StringTok{"PC2"}\NormalTok{, }
       \AttributeTok{title =} \StringTok{"Bray{-}Curtis PCoA"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{(}\DecValTok{12}\NormalTok{) }

\NormalTok{plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{06-beta_files/figure-latex/pcoa_genus-1.pdf}

\hypertarget{estimating-associations-with-an-external-variable}{%
\section{Estimating associations with an external variable}\label{estimating-associations-with-an-external-variable}}

Next to visualizing whether any variable is associated with
differences between samples, we can also quantify the strength of the
association between community composition (beta diversity) and
external factors.

The standard way to do this is to perform a so-called permutational
multivariate analysis of variance (PERMANOVA). This method takes as
input the abundance table, which measure of distance you want to base
the test on and a formula that tells the model how you think the
variables are associated with each other.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# First we get the relative abundance table}
\NormalTok{rel\_abund\_assay }\OtherTok{\textless{}{-}} \FunctionTok{assays}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])}\SpecialCharTok{$}\NormalTok{relabundance}

\CommentTok{\# again transpose it to get taxa to columns}
\NormalTok{rel\_abund\_assay }\OtherTok{\textless{}{-}} \FunctionTok{t}\NormalTok{(rel\_abund\_assay)}

\CommentTok{\# then we can perform the method}
\NormalTok{permanova\_diet }\OtherTok{\textless{}{-}}\NormalTok{ vegan}\SpecialCharTok{::}\FunctionTok{adonis}\NormalTok{(rel\_abund\_assay }\SpecialCharTok{\textasciitilde{}}\NormalTok{ Diet,}
                                  \AttributeTok{data =} \FunctionTok{colData}\NormalTok{(mae),}
                                  \AttributeTok{permutations =} \DecValTok{99}\NormalTok{)}

\CommentTok{\# we can obtain a the p value for our predictor:}
\FunctionTok{print}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{"The test result p{-}value: "}\NormalTok{, }
              \FunctionTok{as.data.frame}\NormalTok{(permanova\_diet}\SpecialCharTok{$}\NormalTok{aov.tab)[}\StringTok{"Diet"}\NormalTok{, }\StringTok{"Pr(\textgreater{}F)"}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "The test result p-value: 0.01"
\end{verbatim}

The diet variable is significantly associated with
microbiota composition (p-value is less than 0.05).

We can visualize those taxa whose abundances drive the
differences between diets. We first need to extract the model
coefficients of taxa:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Gets the coefficients}
\NormalTok{coef }\OtherTok{\textless{}{-}} \FunctionTok{coefficients}\NormalTok{(permanova\_diet)[}\StringTok{"Diet1"}\NormalTok{,]}

\CommentTok{\# Gets the highest coefficients}
\NormalTok{top.coef }\OtherTok{\textless{}{-}} \FunctionTok{sort}\NormalTok{(}\FunctionTok{head}\NormalTok{(coef[}\FunctionTok{rev}\NormalTok{(}\FunctionTok{order}\NormalTok{(}\FunctionTok{abs}\NormalTok{(coef)))],}\DecValTok{20}\NormalTok{))}

\CommentTok{\# Plots the coefficients}
\NormalTok{top\_taxa\_coeffient\_plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\FunctionTok{data.frame}\NormalTok{(}\AttributeTok{x =}\NormalTok{ top.coef,}
                                             \AttributeTok{y =} \FunctionTok{factor}\NormalTok{(}\FunctionTok{names}\NormalTok{(top.coef),}
                         \FunctionTok{unique}\NormalTok{(}\FunctionTok{names}\NormalTok{(top.coef)))),}
                                  \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ x, }\AttributeTok{y =}\NormalTok{ y)) }\SpecialCharTok{+}
  \FunctionTok{geom\_bar}\NormalTok{(}\AttributeTok{stat=}\StringTok{"identity"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{x=}\StringTok{""}\NormalTok{, }\AttributeTok{y=}\StringTok{""}\NormalTok{, }\AttributeTok{title=}\StringTok{"Top Taxa"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{()}

\NormalTok{top\_taxa\_coeffient\_plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{06-beta_files/figure-latex/permanova_coefs-1.pdf}

The above plot shows taxa as code names, and it is hard to tell which
bacterial groups they represent. However, it is easy to add human readable
names. We can fetch those from our rowData. Here we use Genus level names:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Gets corresponding Genus level names and stores them to top.coef}
\NormalTok{names }\OtherTok{\textless{}{-}} \FunctionTok{rowData}\NormalTok{(mae[[}\DecValTok{1}\NormalTok{]])[}\FunctionTok{names}\NormalTok{(top.coef), ][,}\StringTok{"Genus"}\NormalTok{]}

\CommentTok{\# Adds new labels to the plot}
\NormalTok{top\_taxa\_coeffient\_plot }\OtherTok{\textless{}{-}}\NormalTok{ top\_taxa\_coeffient\_plot }\SpecialCharTok{+}
  \FunctionTok{scale\_y\_discrete}\NormalTok{(}\AttributeTok{labels =}\NormalTok{ names) }\CommentTok{\# Adds new labels}
\NormalTok{top\_taxa\_coeffient\_plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{06-beta_files/figure-latex/unnamed-chunk-2-1.pdf}

The same test can be conducted using the ordination from PCoA as follows:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{bray\_curtis\_pcoa\_df}\SpecialCharTok{$}\NormalTok{Diet }\OtherTok{\textless{}{-}} \FunctionTok{colData}\NormalTok{(mae)}\SpecialCharTok{$}\NormalTok{Diet}
\NormalTok{p\_values }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{()}
\ControlFlowTok{for}\NormalTok{(pc }\ControlFlowTok{in} \FunctionTok{c}\NormalTok{(}\StringTok{"pcoa1"}\NormalTok{, }\StringTok{"pcoa2"}\NormalTok{))\{}
  \CommentTok{\# Creates a formula from objects}
\NormalTok{  formula }\OtherTok{\textless{}{-}} \FunctionTok{as.formula}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(pc, }\StringTok{" \textasciitilde{} "}\NormalTok{, }\StringTok{"Diet"}\NormalTok{))}
  \CommentTok{\# Does the permanova analysis}
\NormalTok{  p\_values[[pc]] }\OtherTok{\textless{}{-}}\NormalTok{ vegan}\SpecialCharTok{::}\FunctionTok{adonis}\NormalTok{(formula, }\AttributeTok{data =}\NormalTok{ bray\_curtis\_pcoa\_df,}
                                  \AttributeTok{permutations =} \DecValTok{9999}\NormalTok{, }\AttributeTok{method =} \StringTok{"euclidean"}
\NormalTok{  )}\SpecialCharTok{$}\NormalTok{aov.tab[}\StringTok{"Diet"}\NormalTok{, }\StringTok{"Pr(\textgreater{}F)"}\NormalTok{]}
\NormalTok{\}}

\CommentTok{\# Creates a plot}
\NormalTok{plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ bray\_curtis\_pcoa\_df, }\FunctionTok{aes\_string}\NormalTok{(}\AttributeTok{x =} \StringTok{"pcoa1"}\NormalTok{, }\AttributeTok{y =} \StringTok{"pcoa2"}\NormalTok{, }\AttributeTok{color =} \StringTok{"Diet"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{(}\AttributeTok{size =} \DecValTok{3}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{labs}\NormalTok{(}\AttributeTok{title =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"PCoA beta diversity ordination for microbiome samples"}\NormalTok{), }\AttributeTok{x =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"PC1 (p = "}\NormalTok{, p\_values[[}\StringTok{"pcoa1"}\NormalTok{]], }\StringTok{")"}\NormalTok{), }\AttributeTok{y =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"PC2 (p = "}\NormalTok{, p\_values[[}\StringTok{"pcoa2"}\NormalTok{]], }\StringTok{")"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{theme\_bw}\NormalTok{() }

\NormalTok{plot}
\end{Highlighting}
\end{Shaded}

\includegraphics{06-beta_files/figure-latex/unnamed-chunk-3-1.pdf}

There are many alternative and complementary methods for analysing
community composition. For more examples, see a dedicated section on
beta diversity in the \href{https://microbiome.github.io/OMA/microbiome-diversity.html\#beta-diversity}{online
book}.

\hypertarget{community-typing}{%
\section{Community typing}\label{community-typing}}

A dedicated section presenting examples on community typing is in the
\href{https://microbiome.github.io/OMA/microbiome-community.html\#community-typing}{online book}.

\hypertarget{cross-correlation-unsupervised-learning}{%
\chapter{Cross-correlation \& Unsupervised learning}\label{cross-correlation-unsupervised-learning}}

\hypertarget{cross-correlation}{%
\section{Cross-correlation}\label{cross-correlation}}

With cross-correlation analysis, we can analyze how strongly and how differently variables
are associated between each other. For instance, we can analyze if higher presence of a
specific taxon equals to higher levels of a biomolecule.

\textbf{TASK}
1. Run installation script to load packages into the session
2. Import HintikkaXO data
3. Subset \emph{microbiota} data (rank = ``Phylum'', prevalence = 0.2, detection = 0.001) (\href{https://microbiome.github.io/OMA/differential-abundance.html\#prevalence-filtering}{subsetByPrevalentTaxa})
4. Apply clr-transform to \emph{microbiota} data and log10-tranform to \emph{metabolites} data (\href{https://microbiome.github.io/OMA/taxonomic-information.html\#data-transformation}{transformSamples})
5. Remove uncultured and ambiguous taxa (as it's hard to interpret their results) ( USE THIS: mae{[}{[}1{]}{]} \textless- mae{[}{[}1{]}{]}{[}-grep(``uncultured\textbar Ambiguous\_taxa'', names(mae{[}{[}1{]}{]})),{]} )
5. Calculate cross-correlation between \emph{microbiota} (clr) and \emph{metabolites} (log10) (\href{https://microbiome.github.io/OMA/multi-assay_analyses.html\#multi-assay_analyses}{testExperimentCrossCorrelation})
6. Create a heatmap from cross-correlation matrix \href{https://microbiome.github.io/OMA/microbiome-community.html\#composition-heatmap}{pheatmap}

\hypertarget{unsupervised-learning}{%
\section{Unsupervised learning}\label{unsupervised-learning}}

Unsupervised learning is a part of machine learning where we try to find information
from unknown data. It is also called data mining. Usually this means finding of
clusters, for instance. Cluster refers to group of samples/features that are similar
between each other. For example, based on clinical data we can try to find patient
groups that have similar response to used drug.

\hypertarget{biclustering}{%
\subsection{Biclustering}\label{biclustering}}

Biclustering is a clustering method, which simultaneously clusters rows and columns.
In this example, the aim is to find clusters where subset of taxa share similar
pattern over subset of metabolites. In our case, we try to find clusters where
taxa and metabolites correlate similarly.

Check more from OMA which has dedicated chapter on
\href{https://microbiome.github.io/OMA/biclustering.html}{biclustering}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Load package}
\FunctionTok{library}\NormalTok{(biclust)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: MASS
\end{verbatim}

\begin{verbatim}
## Loading required package: grid
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'grid'
\end{verbatim}

\begin{verbatim}
## The following object is masked from 'package:Biostrings':
## 
##     pattern
\end{verbatim}

\begin{verbatim}
## Loading required package: colorspace
\end{verbatim}

\begin{verbatim}
## Loading required package: lattice
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Find biclusters}
\NormalTok{bc }\OtherTok{\textless{}{-}} \FunctionTok{biclust}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{cor, }\AttributeTok{method=}\FunctionTok{BCPlaid}\NormalTok{(), }\AttributeTok{fit.model =}\NormalTok{ y }\SpecialCharTok{\textasciitilde{}}\NormalTok{ m,}
              \AttributeTok{background =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{shuffle =} \DecValTok{100}\NormalTok{, }\AttributeTok{back.fit =} \DecValTok{0}\NormalTok{, }\AttributeTok{max.layers =} \DecValTok{10}\NormalTok{,}
              \AttributeTok{iter.startup =} \DecValTok{10}\NormalTok{, }\AttributeTok{iter.layer =} \DecValTok{100}\NormalTok{, }\AttributeTok{verbose =} \ConstantTok{FALSE}\NormalTok{)}

\NormalTok{bc}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## An object of class Biclust 
## 
## call:
##  biclust(x = corr$cor, method = BCPlaid(), fit.model = y ~ m, 
##      background = TRUE, shuffle = 100, back.fit = 0, max.layers = 10, 
##      iter.startup = 10, iter.layer = 100, verbose = FALSE)
## 
## There was one cluster found with
##   4 Rows and  6 columns
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Get biclusters}
\NormalTok{bicluster\_rows }\OtherTok{\textless{}{-}}\NormalTok{ bc}\SpecialCharTok{@}\NormalTok{RowxNumber}
\NormalTok{bicluster\_columns }\OtherTok{\textless{}{-}}\NormalTok{ bc}\SpecialCharTok{@}\NormalTok{NumberxCol}

\CommentTok{\# Convert them into data.frames}
\NormalTok{bicluster\_rows }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(bicluster\_rows)}
\NormalTok{bicluster\_columns }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{t}\NormalTok{(bc}\SpecialCharTok{@}\NormalTok{NumberxCol))}

\CommentTok{\# Adjust names of clusters}
\FunctionTok{colnames}\NormalTok{(bicluster\_rows) }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"cluster\_"}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(bicluster\_rows))}
\FunctionTok{colnames}\NormalTok{(bicluster\_columns) }\OtherTok{\textless{}{-}} \FunctionTok{paste0}\NormalTok{(}\StringTok{"cluster\_"}\NormalTok{, }\DecValTok{1}\SpecialCharTok{:}\FunctionTok{ncol}\NormalTok{(bicluster\_columns))}

\CommentTok{\# Print biclusters for rows}
\FunctionTok{head}\NormalTok{(bicluster\_rows)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   cluster_1
## 1     FALSE
## 2     FALSE
## 3     FALSE
## 4     FALSE
## 5     FALSE
## 6     FALSE
\end{verbatim}

Now, we can add bicluster information into the heatmap that we already made.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Convert boolean values into factors}
\NormalTok{bicluster\_columns[ , }\DecValTok{1}\NormalTok{] }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(bicluster\_columns[, }\DecValTok{1}\NormalTok{] )}
\NormalTok{bicluster\_rows[ , }\DecValTok{1}\NormalTok{] }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(bicluster\_rows[ , }\DecValTok{1}\NormalTok{])}

\CommentTok{\# Adjust their rownames}
\FunctionTok{rownames}\NormalTok{(bicluster\_columns) }\OtherTok{\textless{}{-}} \FunctionTok{colnames}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{cor)}
\FunctionTok{rownames}\NormalTok{(bicluster\_rows) }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{cor)}

\CommentTok{\# Get correlation values that are over thresholds}
\NormalTok{p\_threshold }\OtherTok{\textless{}{-}} \FloatTok{0.01}
\NormalTok{corr\_values }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{p\_adj}\SpecialCharTok{\textless{}}\NormalTok{p\_threshold, }\FunctionTok{round}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{cor,}\DecValTok{1}\NormalTok{), }\StringTok{""}\NormalTok{)}

\CommentTok{\# Create a heatmap}
\FunctionTok{pheatmap}\NormalTok{(corr}\SpecialCharTok{$}\NormalTok{cor,}
         \AttributeTok{annotation\_col =}\NormalTok{ bicluster\_columns, }
         \AttributeTok{annotation\_row =}\NormalTok{ bicluster\_rows,}
         
         \AttributeTok{display\_numbers =}\NormalTok{ corr\_values,}
   
         \AttributeTok{main =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"Correlations between bacteria and metabolites }
\StringTok{                       (correlation over threshold (p \textless{} "}\NormalTok{, p\_threshold,}\StringTok{") marked)"}\NormalTok{),}
         
         \AttributeTok{breaks =}\NormalTok{ breaks,}
         \AttributeTok{color =}\NormalTok{ colors, }
   
         \AttributeTok{fontsize\_number =} \DecValTok{6}\NormalTok{, }
         \AttributeTok{number\_color =} \StringTok{"yellow"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{07-unsupML_files/figure-latex/unnamed-chunk-11-1.pdf}

\hypertarget{supervised-learning}{%
\chapter{Supervised learning}\label{supervised-learning}}

Machine learning models are highly flexible and can be used to model differences in
samples, similarly to (frequentist) statistics. However, the analysis workflow with
these methods is very different from the frequentist analyses. These models learn
a function to predict values of the dependent variable, given data. Between different
models, the algorithms vary greatly, but generally all regression and classification
models can be used similarly in a machine learning workflow.

Machine learning models do not usually output p-values, but they are designed to
predict the outcome (value or class) of the dependent variable based on data.
Thus, if we want to know how good our model is, we need to divide our data to training
and test (or validation) sets. The training set is used to train the model, and the
validation set can be then used to test the model. The model can be used to predict
the outcome of the dependent variable on the test data, and the predicted values can
be compared to the actual known values in the test data. It is important to make sure
that there is no data leakage between these two sets, or otherwise the validation is
compromised.

In the workshop we use random forests and the caret package to train regression and
classification models. The models will predict continuous butyrate concentration or discretized
class (high/low butyrate) based on the microbiome composition (\href{https://scholar.google.com/scholar?as_sdt=0\%2C5\&as_ylo=2015\&q=butyrate+and+gut+microbiome\&btnG=}{why butyrate?}).

\hypertarget{data-curation}{%
\section{Data curation}\label{data-curation}}

We first make a data frame which includes only the butyrate concentration and the
transformed genus-level microbiome data.

Creating a dataframe for modeling butyrate levels:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{butyrate\_df }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\FunctionTok{cbind}\NormalTok{(y, x))}
\NormalTok{butyrate\_df }\OtherTok{\textless{}{-}}\NormalTok{ butyrate\_df[,}\FunctionTok{which}\NormalTok{(}\FunctionTok{colnames}\NormalTok{(butyrate\_df) }\SpecialCharTok{\%in\%} \FunctionTok{c}\NormalTok{(}\StringTok{"Butyrate"}\NormalTok{, }\FunctionTok{colnames}\NormalTok{(x)))]}
\end{Highlighting}
\end{Shaded}

A function in the \href{https://topepo.github.io/caret/}{caret package} is used to divide the data once to 80\% train and 20\% test (validation)
sets. This is to prevent data leakage and overestimation of model performance. The
20\% test set is only used to conduct the final validation of the models. The number of
samples in this case is low (n = 40), but as we can see later on, even 8 samples is
sufficient for estimation of the performance. Note that the data is stratified to include
a representative distribution of butyrate concentrations on both sides of the split.
There is some randomness inherent in the splitting, so set.seed() needs to be used.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(caret)}
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{42}\NormalTok{)}
\NormalTok{trainIndex }\OtherTok{\textless{}{-}} \FunctionTok{createDataPartition}\NormalTok{(butyrate\_df}\SpecialCharTok{$}\NormalTok{Butyrate, }\AttributeTok{p =}\NormalTok{ .}\DecValTok{8}\NormalTok{, }\AttributeTok{list =} \ConstantTok{FALSE}\NormalTok{, }\AttributeTok{times =} \DecValTok{1}\NormalTok{)}
\NormalTok{butyrate\_df\_train }\OtherTok{\textless{}{-}}\NormalTok{ butyrate\_df[trainIndex,]}
\NormalTok{butyrate\_df\_test }\OtherTok{\textless{}{-}}\NormalTok{ butyrate\_df[}\SpecialCharTok{{-}}\NormalTok{trainIndex,]}
\end{Highlighting}
\end{Shaded}

\hypertarget{regression-with-random-forests}{%
\section{Regression with random forests}\label{regression-with-random-forests}}

Random forests are a common and flexible ensemble learning method, which are a good starting
point when choosing a machine learning model. We are using the \href{https://github.com/imbs-hl/ranger}{ranger} implementation
of random forests, which runs quite fast compared to its alternatives in R. A wrapper
train function from caret is used to conduct a 5-fold cross-validation, repeated 5 times
with random partitions inside the training data (for further reading, \href{https://topepo.github.io/caret/model-training-and-tuning.html}{see documentation}).
Because of the randomness, seed needs to be set again. \textbf{Note that the seed does \emph{not} remain set if you re-run a function with a random component
without calling set.seed() first!} Specific to random forests, we are also using an option to use permutation
importance.

The train function is \href{https://www.rdocumentation.org/packages/caret/versions/4.47/topics/train}{quite complex}, and performs hyperparameter tuning while training the
model with cross-validation. The final model included in the object is then trained on all
input data and optimized hyperparameters.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{42}\NormalTok{)}
\NormalTok{fitControl }\OtherTok{\textless{}{-}} \FunctionTok{trainControl}\NormalTok{(}\AttributeTok{method =} \StringTok{"repeatedcv"}\NormalTok{, }\AttributeTok{number =} \DecValTok{5}\NormalTok{, }\AttributeTok{repeats =} \DecValTok{5}\NormalTok{)}
\NormalTok{rfFit1 }\OtherTok{\textless{}{-}} \FunctionTok{train}\NormalTok{(Butyrate }\SpecialCharTok{\textasciitilde{}}\NormalTok{ ., }\AttributeTok{data =}\NormalTok{ butyrate\_df\_train, }
                \AttributeTok{method =} \StringTok{"ranger"}\NormalTok{, }
                \AttributeTok{trControl =}\NormalTok{ fitControl,}
                \AttributeTok{importance =} \StringTok{"permutation"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Following the training, we can print out the resulting object, which shows details about
the training. We also print out results of the final model.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(rfFit1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Random Forest 
## 
## 32 samples
## 28 predictors
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold, repeated 5 times) 
## Summary of sample sizes: 24, 26, 25, 28, 25, 25, ... 
## Resampling results across tuning parameters:
## 
##   mtry  splitrule   RMSE       Rsquared   MAE      
##    2    variance    0.8801800  0.5870278  0.6802341
##    2    extratrees  0.9039197  0.5888711  0.7113279
##   15    variance    0.8604439  0.5806824  0.6573050
##   15    extratrees  0.8678618  0.5773606  0.6682391
##   28    variance    0.8702166  0.5663926  0.6564932
##   28    extratrees  0.8569955  0.5883242  0.6589236
## 
## Tuning parameter 'min.node.size' was held constant at a value of 5
## RMSE was used to select the optimal model using the smallest value.
## The final values used for the model were mtry = 28, splitrule = extratrees
##  and min.node.size = 5.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(rfFit1}\SpecialCharTok{$}\NormalTok{finalModel)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Ranger result
## 
## Call:
##  ranger::ranger(dependent.variable.name = ".outcome", data = x,      mtry = min(param$mtry, ncol(x)), min.node.size = param$min.node.size,      splitrule = as.character(param$splitrule), write.forest = TRUE,      probability = classProbs, ...) 
## 
## Type:                             Regression 
## Number of trees:                  500 
## Sample size:                      32 
## Number of independent variables:  28 
## Mtry:                             28 
## Target node size:                 5 
## Variable importance mode:         permutation 
## Splitrule:                        extratrees 
## Number of random splits:          1 
## OOB prediction error (MSE):       0.8650469 
## R squared (OOB):                  0.4563803
\end{verbatim}

We can then compare the metrics produced in training to actual validation metrics. Here,
the final model predicts on unseen test data samples, and metrics are calculated
against the observed (true) values. We can also plot the predicted values against the
observed values, and include a line to show how a perfect model would predict.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test\_predictions }\OtherTok{\textless{}{-}} \FunctionTok{predict}\NormalTok{(rfFit1, }\AttributeTok{newdata =}\NormalTok{ butyrate\_df\_test)}
\FunctionTok{print}\NormalTok{(}\FunctionTok{postResample}\NormalTok{(test\_predictions, butyrate\_df\_test}\SpecialCharTok{$}\NormalTok{Butyrate))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      RMSE  Rsquared       MAE 
## 1.1129878 0.4973854 0.9576183
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Plot predicted vs observed}
\NormalTok{pred\_obs }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{predicted =}\NormalTok{ test\_predictions, }\AttributeTok{observed =}\NormalTok{ butyrate\_df\_test}\SpecialCharTok{$}\NormalTok{Butyrate)}
\FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ pred\_obs, }\FunctionTok{aes}\NormalTok{(}\AttributeTok{x=}\NormalTok{predicted, }\AttributeTok{y=}\NormalTok{observed)) }\SpecialCharTok{+} \FunctionTok{geom\_point}\NormalTok{(}\AttributeTok{size =} \DecValTok{5}\NormalTok{, }\AttributeTok{color =} \StringTok{"orange"}\NormalTok{) }\SpecialCharTok{+} 
  \FunctionTok{xlab}\NormalTok{(}\StringTok{"Predicted butyrate concentration"}\NormalTok{) }\SpecialCharTok{+} \FunctionTok{ylab}\NormalTok{(}\StringTok{"Observed butyrate concentration"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{lims}\NormalTok{(}\AttributeTok{x =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{5}\NormalTok{), }\AttributeTok{y =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{5}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{geom\_abline}\NormalTok{(}\AttributeTok{linetype =} \DecValTok{5}\NormalTok{, }\AttributeTok{color =} \StringTok{"blue"}\NormalTok{, }\AttributeTok{size =} \DecValTok{1}\NormalTok{)}\SpecialCharTok{+} \CommentTok{\# Plot a perfect fit line}
    \FunctionTok{theme}\NormalTok{(}\AttributeTok{panel.border =} \FunctionTok{element\_rect}\NormalTok{(}\AttributeTok{colour =} \StringTok{"black"}\NormalTok{, }\AttributeTok{fill =} \ConstantTok{NA}\NormalTok{),}
                             \AttributeTok{panel.background =} \FunctionTok{element\_blank}\NormalTok{())}
\end{Highlighting}
\end{Shaded}

\includegraphics{08-supML_files/figure-latex/unnamed-chunk-6-1.pdf}

All models are wrong, but some are useful. Luckily, our model seems to be somewhat useful.
However, we are not yet done with the model. It is important to examine how our model
actually works. Even though the random forest can look like a ``black box'', we can get
much information out on how it ends up making specific predictions.

Feature importance can be used, quite literally, to see which features are important
for model predictions. Permutation importance is a metric calculated by shuffling the
values of individual feature columns. Permutation importance of a feature will be high
if such corrupted data leads to bad predictions of the model. If shuffling a feature
does not affect the model performance negatively, its permutation importance will be low.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(}\FunctionTok{varImp}\NormalTok{(rfFit1))}
\end{Highlighting}
\end{Shaded}

\includegraphics{08-supML_files/figure-latex/unnamed-chunk-7-1.pdf}

We can see that some genera are highly important for model predictions. The importance
values can be highly useful, and these are often used \emph{e.g.,} for feature selection before
conducting statistical (or other ML) tests. However, if we only conducted this supervised
machine learning analysis, we would not know which features are positively and which
ones negatively associated with butyrate levels.

All black box models can however be examined through partial dependence plots. Similarly to
validation, we can again utilize the fact that ML models are great at making new predictions.
While the inner workings of the model are highly complex, we can assume that changing
the values of an important feature affects the model prediction in a specific way.
Briefly, partial dependence plots visualize the expected output of the model over the
range of an individual input feature (up to 3 features). The \href{https://bgreenwell.github.io/pdp/articles/pdp.html}{pdp package}
is a versatile implementation for conducting these analyses in R and works directly on models
fitted with caret::train().

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(patchwork)}
\FunctionTok{library}\NormalTok{(pdp)}
\NormalTok{top\_features }\OtherTok{\textless{}{-}} \FunctionTok{rownames}\NormalTok{(}\FunctionTok{varImp}\NormalTok{(rfFit1)}\SpecialCharTok{$}\NormalTok{importance)[}\FunctionTok{order}\NormalTok{(}\FunctionTok{varImp}\NormalTok{(rfFit1)}\SpecialCharTok{$}\NormalTok{importance[,}\StringTok{"Overall"}\NormalTok{], }\AttributeTok{decreasing =} \ConstantTok{TRUE}\NormalTok{)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{6}\NormalTok{]]}
\NormalTok{pd\_plots }\OtherTok{\textless{}{-}} \FunctionTok{list}\NormalTok{(}\ConstantTok{NULL}\NormalTok{)}
\ControlFlowTok{for}\NormalTok{ (feature }\ControlFlowTok{in} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{length}\NormalTok{(top\_features)) \{}
\NormalTok{  pd\_plots[[feature]] }\OtherTok{\textless{}{-}} \FunctionTok{partial}\NormalTok{(rfFit1, }\AttributeTok{pred.var =}\NormalTok{ top\_features[feature], }\AttributeTok{rug =} \ConstantTok{TRUE}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{autoplot}\NormalTok{() }\SpecialCharTok{+} 
    \FunctionTok{geom\_hline}\NormalTok{(}\AttributeTok{yintercept =} \FunctionTok{mean}\NormalTok{(butyrate\_df\_train}\SpecialCharTok{$}\NormalTok{Butyrate), }\AttributeTok{linetype =} \DecValTok{2}\NormalTok{, }\AttributeTok{color =} \StringTok{"gray"}\NormalTok{) }\SpecialCharTok{+} \CommentTok{\# Show the mean of the training data as a dashed line}
    \FunctionTok{scale\_y\_continuous}\NormalTok{(}\AttributeTok{limits=}\FunctionTok{c}\NormalTok{(}\FloatTok{1.5}\NormalTok{,}\FloatTok{2.3}\NormalTok{)) }\SpecialCharTok{+} \CommentTok{\# Harmonize the scale of yhat on all plots}
      \FunctionTok{theme}\NormalTok{(}\AttributeTok{panel.border =} \FunctionTok{element\_rect}\NormalTok{(}\AttributeTok{colour =} \StringTok{"black"}\NormalTok{, }\AttributeTok{fill =} \ConstantTok{NA}\NormalTok{),}
                             \AttributeTok{panel.background =} \FunctionTok{element\_blank}\NormalTok{())}
  \FunctionTok{print}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(}\StringTok{"Partial dependence of "}\NormalTok{, top\_features[feature]))}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Partial dependence of UBA1819"
## [1] "Partial dependence of Faecalitalea"
## [1] "Partial dependence of Papillibacter"
## [1] "Partial dependence of Anaerostipes"
## [1] "Partial dependence of Lachnoclostridium"
## [1] "Partial dependence of Ruminiclostridium"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{wrap\_plots}\NormalTok{(pd\_plots)}
\end{Highlighting}
\end{Shaded}

\includegraphics{08-supML_files/figure-latex/unnamed-chunk-8-1.pdf}

\hypertarget{classification-with-random-forests}{%
\section{Classification with random forests}\label{classification-with-random-forests}}

In addition to regression, random forests can also be used for classification tasks.
While the actual butyrate values might be important for some purposes, it can be
expected that it is quite hard to model the concentration throughout its observed
range accurately. As a demonstration, we can also discretize a constant variable,
and handle the problem as a classification task. Here, we are using the median value
of butyrate as a cutoff for high and low butyrate groups - which makes this a binary
classification. Other ways to specify a cutoff for discretization are often more
justified than just the mean or median. These are usually informed by previous studies
and results.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{butyrate\_cutoff }\OtherTok{\textless{}{-}} \FunctionTok{median}\NormalTok{(butyrate\_df\_test}\SpecialCharTok{$}\NormalTok{Butyrate)}
\NormalTok{butyrate\_df\_test\_2 }\OtherTok{\textless{}{-}}\NormalTok{ butyrate\_df\_test}
\NormalTok{butyrate\_df\_train\_2 }\OtherTok{\textless{}{-}}\NormalTok{ butyrate\_df\_train}
\NormalTok{butyrate\_df\_test\_2}\SpecialCharTok{$}\NormalTok{Butyrate }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(}\FunctionTok{ifelse}\NormalTok{(butyrate\_df\_test\_2}\SpecialCharTok{$}\NormalTok{Butyrate }\SpecialCharTok{\textgreater{}=}\NormalTok{ butyrate\_cutoff, }\StringTok{"High"}\NormalTok{, }\StringTok{"Low"}\NormalTok{))}
\NormalTok{butyrate\_df\_train\_2}\SpecialCharTok{$}\NormalTok{Butyrate }\OtherTok{\textless{}{-}} \FunctionTok{as.factor}\NormalTok{(}\FunctionTok{ifelse}\NormalTok{(butyrate\_df\_train\_2}\SpecialCharTok{$}\NormalTok{Butyrate }\SpecialCharTok{\textgreater{}=}\NormalTok{ butyrate\_cutoff, }\StringTok{"High"}\NormalTok{, }\StringTok{"Low"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Training of the model is very similar to regression, but we want to define two options in the
trainControl function sent to caret::train(). classProbs = TRUE is used to output classification
probabilities instead of the classes themselves. This is required for calculating
ROC-AUC values. Also, we are defining a summaryFunction which is used for evaluation
and hyperparameter optimization.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{set.seed}\NormalTok{(}\DecValTok{42}\NormalTok{)}
\NormalTok{fitControl }\OtherTok{\textless{}{-}} \FunctionTok{trainControl}\NormalTok{(}\AttributeTok{method =} \StringTok{"repeatedcv"}\NormalTok{, }\AttributeTok{number =} \DecValTok{5}\NormalTok{, }\AttributeTok{repeats =} \DecValTok{10}\NormalTok{, }\AttributeTok{classProbs =} \ConstantTok{TRUE}\NormalTok{, }\AttributeTok{summaryFunction =}\NormalTok{ twoClassSummary)}
\NormalTok{rfFit2 }\OtherTok{\textless{}{-}} \FunctionTok{train}\NormalTok{(Butyrate }\SpecialCharTok{\textasciitilde{}}\NormalTok{ ., }\AttributeTok{data =}\NormalTok{ butyrate\_df\_train\_2, }
                \AttributeTok{method =} \StringTok{"ranger"}\NormalTok{, }
                \AttributeTok{trControl =}\NormalTok{ fitControl,}
                \AttributeTok{importance =} \StringTok{"permutation"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We then measure performance with test data. Here, we need to construct a data frame with
specific dimensions as an input for twoClassSummary().

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{test\_predictions\_2 }\OtherTok{\textless{}{-}} \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{obs =}\NormalTok{ butyrate\_df\_test\_2}\SpecialCharTok{$}\NormalTok{Butyrate, }
                                 \AttributeTok{pred =} \FunctionTok{predict}\NormalTok{(rfFit2, }\AttributeTok{newdata =}\NormalTok{ butyrate\_df\_test\_2), }
                                 \FunctionTok{predict}\NormalTok{(rfFit2, }\AttributeTok{newdata =}\NormalTok{ butyrate\_df\_test\_2, }\AttributeTok{type =} \StringTok{"prob"}\NormalTok{))}

\CommentTok{\#Print out the metrics}
\FunctionTok{print}\NormalTok{(rfFit2)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Random Forest 
## 
## 32 samples
## 28 predictors
##  2 classes: 'High', 'Low' 
## 
## No pre-processing
## Resampling: Cross-Validated (5 fold, repeated 10 times) 
## Summary of sample sizes: 25, 25, 26, 26, 26, 26, ... 
## Resampling results across tuning parameters:
## 
##   mtry  splitrule   ROC        Sens       Spec     
##    2    gini        0.8622222  0.7733333  0.8433333
##    2    extratrees  0.8877778  0.8133333  0.8450000
##   15    gini        0.8261111  0.7866667  0.7950000
##   15    extratrees  0.8561111  0.8266667  0.8116667
##   28    gini        0.8150000  0.7733333  0.7950000
##   28    extratrees  0.8488889  0.8200000  0.8016667
## 
## Tuning parameter 'min.node.size' was held constant at a value of 1
## ROC was used to select the optimal model using the largest value.
## The final values used for the model were mtry = 2, splitrule = extratrees
##  and min.node.size = 1.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(rfFit2}\SpecialCharTok{$}\NormalTok{finalModel)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Ranger result
## 
## Call:
##  ranger::ranger(dependent.variable.name = ".outcome", data = x,      mtry = min(param$mtry, ncol(x)), min.node.size = param$min.node.size,      splitrule = as.character(param$splitrule), write.forest = TRUE,      probability = classProbs, ...) 
## 
## Type:                             Probability estimation 
## Number of trees:                  500 
## Sample size:                      32 
## Number of independent variables:  28 
## Mtry:                             2 
## Target node size:                 1 
## Variable importance mode:         permutation 
## Splitrule:                        extratrees 
## Number of random splits:          1 
## OOB prediction error (Brier s.):  0.163163
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\FunctionTok{twoClassSummary}\NormalTok{(test\_predictions\_2, }\AttributeTok{lev =} \FunctionTok{c}\NormalTok{(}\StringTok{"High"}\NormalTok{, }\StringTok{"Low"}\NormalTok{)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   ROC  Sens  Spec 
## 0.875 0.750 0.750
\end{verbatim}

Often just the ROC-AUC value calculated above can suffice, for example for
model comparisons. However, if class distribution is skewed, area under the precision-recall curve
(AUPRC) should be used instead (see \href{https://doi.org/10.1002/bimj.201800148}{Fu et al., 2018}).

Here is one way to calculate both with the package precrec and plot the curves.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(MLmetrics)}
\FunctionTok{library}\NormalTok{(precrec)}
\NormalTok{aucs }\OtherTok{\textless{}{-}} \FunctionTok{evalmod}\NormalTok{(}\AttributeTok{scores =}\NormalTok{ test\_predictions\_2}\SpecialCharTok{$}\NormalTok{Low, }\AttributeTok{labels =}\NormalTok{ test\_predictions\_2}\SpecialCharTok{$}\NormalTok{obs)}
\FunctionTok{print}\NormalTok{(aucs)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
##     === AUCs ===
## 
##      Model name Dataset ID Curve type       AUC
##    1         m1          1        ROC 0.8750000
##    2         m1          1        PRC 0.8722936
## 
## 
##     === Input data ===
## 
##      Model name Dataset ID # of negatives # of positives
##    1         m1          1              4              4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{autoplot}\NormalTok{(aucs)}
\end{Highlighting}
\end{Shaded}

\includegraphics{08-supML_files/figure-latex/unnamed-chunk-12-1.pdf}

Finally, we can also extract and plot the feature importance from the binary
classification model. It is interesting to compare this result to the importances
of the same features in the regression model.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot}\NormalTok{(}\FunctionTok{varImp}\NormalTok{(rfFit2))}
\end{Highlighting}
\end{Shaded}

\includegraphics{08-supML_files/figure-latex/unnamed-chunk-13-1.pdf}

  \bibliography{packages.bib}

\end{document}
